src/Optical/LensAssembly.jl

```diff
@@ -170,55 +126,47 @@ macro lensassembly_elements(N)
     end)
 end
 
-# pregenerate a small number at compile time
-for N in 1:PREGENERATED_LENS_ASSEMBLY_SIZE
-    type = :($(Symbol("LensAssembly$(N)")))
-    if !isdefined(OpticSim, type)
+function compile_lensassembly(N)
+    if !isdefined(OpticSim, :($(Symbol("LensAssembly$N"))))
         @eval @lensassembly_constructor($N)
         @eval @lensassembly_intersection($N)
         @eval @lensassembly_elements($N)
     end
 end
 
-function LensAssembly(elements::Vararg{Union{Surface{T},CSGTree{T},LensAssembly{T}}}; axis::SVector{3,T} = SVector{3,T}(0.0, 0.0, 1.0)) where {T<:Real}
+# pregenerate a small number at compile time
+compile_lensassembly.(1:PREGENERATED_LENS_ASSEMBLY_SIZE)
+
+function LensAssembly(elements::Vararg{LensAssemblyElement{T}}; axis::SVector{3,T} = SVector{3,T}(0.0, 0.0, 1.0)) where {T<:Real}
     # make the actual object
-    actual_elements = []
-    rectangles = Vector{Rectangle{T}}(undef, 0)
```

Comment:
I think I did this originally to avoid abstract types and consequent allocations - maybe this isn't relevant anymore?
The typed elements get explicitly parameterized in the LensAssembly type so are ok, but here I would think the array of `PlanarShape{T}` (which is abstract) results in tracing objects of unspecified type in `closest_intersection(lens_assembly, ray)`

Reply:
It seems to pass the tests without allocations, so maybe it's ok now? Otherwise, are you saying we'll have to revert back to separate rectangle/ellipse/hexagon vectors?

Reply:
Reading [this](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container), it seems that this change may or may not cause allocations, but it's at least less memory efficient. Is this an issue? Or are we happy as long as there are no allocations?

Reply:
Runtime allocations (at least used to) result in performance that was many orders of magnitude slower, so providing allocs stay at  you're probably ok even if the code seems less efficient. You can always look at the benchmark time as well as the allocs as a sanity check

Reply:
```
Allocations: Error During Test at /home/runner/work/OpticSim.jl/OpticSim.jl/test/testsets/Allocations.jl:8
  Expression evaluated to non-Boolean
  Expression: (Benchmarks.runbenchmark(b, samples = 1, evals = 1)).allocs == 0 || "$(b)"
       Value: "planar_shapes"
```
😢

Reply:
> Do you remember if there are tests targeting allocs on rectangle/ellipse/hexagon-based LensAssemblies?

Really great catch on this - narrowly avoided a major performance regression due to my oversight when writing the alloc tests initially!

Reply:
:'(

At least we know the test works!

Reply:
Yeah this sounds like a sensible idea - it seems redundant to have them all as separate implementations in ConvexPolygon can do the same thing. I guess Ellipse may have a purpose still, also would be interested in performance comparison between e.g. Rectangle and ConvecPolygon - though I expect there will be little difference